### PR TITLE
DOC/ERR: better error message on no common merge keys

### LINKED ISSUE #19391: DOC/ERR: better error message on no common merge keys
Currently, merging two DataFrames on indices is not a default option. When the keyword argument `on` is not specified, `MergeError` is occurred.

#### Code Sample and Problem Statement

```python
In [1]: import pandas as pd
In [2]: a = pd.DataFrame({'a':[1,2]})
In [3]: b = pd.DataFrame({'b':[10,20]})
In [4]: a.merge(b)
---------------------------------------------------------------------------
MergeError                                Traceback (most recent call last)
<ipython-input-6-fbb650d95f99> in <module>()
----> 1 a.merge(b)

/home/swyoon/env/swyoon/local/lib/python2.7/site-packages/pandas/core/frame.pyc in merge(self, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)
   5368                      right_on=right_on, left_index=left_index,
   5369                      right_index=right_index, sort=sort, suffixes=suffixes,
-> 5370                      copy=copy, indicator=indicator, validate=validate)
   5371
   5372     def round(self, decimals=0, *args, **kwargs):

/home/swyoon/env/swyoon/local/lib/python2.7/site-packages/pandas/core/reshape/merge.pyc in merge(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)
     55                          right_index=right_index, sort=sort, suffixes=suffixes,
     56                          copy=copy, indicator=indicator,
---> 57                          validate=validate)
     58     return op.get_result()
     59

/home/swyoon/env/swyoon/local/lib/python2.7/site-packages/pandas/core/reshape/merge.pyc in __init__(self, left, right, how, on, left_on, right_on, axis, left_index, right_index, sort, suffixes, copy, indicator, validate)
    558             warnings.warn(msg, UserWarning)
    559
--> 560         self._validate_specification()
    561
    562         # note this function has side effects

/home/swyoon/env/swyoon/local/lib/python2.7/site-packages/pandas/core/reshape/merge.pyc in _validate_specification(self)
    951                     self.right.columns)
    952                 if len(common_cols) == 0:
--> 953                     raise MergeError('No common columns to perform merge on')
    954                 if not common_cols.is_unique:
    955                     raise MergeError("Data columns not unique: {common!r}"

MergeError: No common columns to perform merge on
```

For a successful merge, we need to specify rather verbose keyword arguments `left_index=True, right_index=True`.

```python
In [5]: a.merge(b, left_index=True, right_index=True)
Out[5]:
   a   b
0  1  10
1  2  20
```

#### Expected Output

What if we could do

```python
In [1]: import pandas as pd
In [2]: a = pd.DataFrame({'a':[1,2]})
In [3]: b = pd.DataFrame({'b':[10,20]})
In [4]: a.merge(b)
Out[4]:
   a   b
0  1  10
1  2  20
```

A briefer syntax for merge on index could facilitate the usability.

#### Output of ``pd.show_versions()``

<details>

```python
In [5]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.12.final.0
python-bits: 64
OS: Linux
OS-release: 4.2.0-42-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.22.0
pytest: 3.1.3
pip: 9.0.1
setuptools: 38.2.4
Cython: None
numpy: 1.14.0
scipy: 0.19.1
pyarrow: None
xarray: None
IPython: 5.4.1
sphinx: None
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: 2.4.8
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml: 3.7.3
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: 1.1.15
pymysql: None
psycopg2: 2.7.1 (dt dec pq3 ext lo64)
jinja2: 2.9.6
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
```
</details>

### PR DESCRIPTION
**Pull request by swyoon**
- [o] closes #19391 
- [o] tests passed
- [o] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- Let MergeError emit values keyword arguments
- Add `DataFrame.join` on `See also` section of  `pandas.merge`

### CODE REVIEW

**Comment by swyoon** [comment]
It seems that the failed test has nothing to to with the content of the commit though..

**Comment by TomAugspurger** [comment]
Can you add a test for this, and a release note?

What's the output look like for large DataFrames? Does the repr truncation work correctly?

File: pandas/core/frame.py [hunk 1]
```diff
@@ -233,7 +233,7 @@
 --------
 merge_ordered
 merge_asof
-
+DataFrame.join : For index-based merge operations
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: I think you can remove the explanation, everything other than the name. I'm not sure what numpydoc does with it, but those are turned into hyperlinks for the HTML docs.

**Comment by swyoon** [comment]
Release note added, `see also` modified
I believe my contribution is about the error message. Should the error message be included in the test?
It seems that there already exists a test for [a merge with no common keys](https://github.com/pandas-dev/pandas/blob/fb3b23782534c925ab7158c59dcb32c8f8390d71/pandas/tests/reshape/merge/test_merge.py#L266).
I don't get what you said about the large DataFrames. Could you explain a bit more?

**Comment by TomAugspurger** [comment]
> I believe my contribution is about the error message. Should the error message be included in the test?

Yes, see `pandas.util.testing.assert_raises_regex` for how to do it. You'll want to do a few different tests and assert that the correct information is always in the exception message.

Ignore my comment about large dataframes I think.

File: pandas/core/reshape/merge.py [hunk 2]
```diff
@@ -1028,7 +1028,14 @@ def _validate_specification(self):
                 common_cols = self.left.columns.intersection(
                     self.right.columns)
                 if len(common_cols) == 0:
-                    raise MergeError('No common columns to perform merge on')
+                    raise MergeError('No common columns to perform merge on. '
+                                     'Merge options: right_on={ron}, '
```
Comment (TomAugspurger) [inline pandas/core/reshape/merge.py]: Can you change the order to be `left_on, right_on, left_index, right_index`, to match the order in the signature?

Comment (swyoon) [reply pandas/core/reshape/merge.py]: Oh sure I can. will do it.

Comment (jreback) [reply pandas/core/reshape/merge.py]: to make this more readable you can break it like
```
raise MergeError(
      'No common ......'   

)
```

File: pandas/core/reshape/merge.py [hunk 3]
```diff
@@ -1028,7 +1028,14 @@ def _validate_specification(self):
...
+                                     'left_on={lon}, right_index={ridx}, '
+                                     'left_index={lidx}'
+                                     .format(ron=self.right_on,
+                                             lon=self.left_on,
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: can you update the test for this (it prob just raises) and match the error message with ``assert_raises_regex``

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you rebase and update

**Comment by swyoon** [comment]
sure, i'm almost done. i will update in a few minutes.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @swyoon

**Comment by swyoon** [comment]
thanks for kind reviews! @jreback @TomAugspurger

**Pull request merged**